### PR TITLE
Streaming support (spike)

### PR DESCRIPTION
**Pull request by AndrewRadev**
### Describe the change

The idea is to make it possible for a pager to continuously get output and render it. With this change, the system pager is able to do this. Here's an example that works for me (run in the root of the project):

``` ruby
$: << File.expand_path('./lib')
require 'tty-pager'

pager = TTY::Pager.new
file = File.join(File.dirname(__FILE__), 'examples/temp.txt')

File.open(file).each_line do |line|
  sleep 0.1
  if !pager.write(line)
    STDERR.puts "Pager was closed"
    break
  end
end

pager.wait
```

There are many more things to do and questions to consider:

### Why are we doing this? Benefits & Drawbacks

See discussion in https://github.com/piotrmurach/tty-pager/issues/13

### Requirements
Put an X between brackets on each line if you have done the item:
[x] Tests written & passing locally?
[] Code style checked? -- HoundCI seems over-eager, so not sure about code style. Happy to change.
[x] Rebased with `master` branch?
[x] Documentaion updated?

### CODE REVIEW

**Comment by piotrmurach** [comment]
Thanks so much for working on this! Great to see progress on possible API. Unfortunately, I haven't had time to dig into your solution yet. I will give it a try over the weekend.

One thing I was thinking is, for example, how would you page a large file through the new API?

In other words, it feels to me that there is a common abstraction that links the `BasicPager` and `SystemPager` together. Maybe, an enumerable like object that works on newline basis. If you give it a file descriptor it will 'know' how to feed lines to both pagers. Similarly, if you write a line, it will consume it and yield back to the pager. I'm not sure if I'm making much sense. The whole thing is fuzzy to me and I need to spend some time playing with it like yourself. Hopefully I'll get some time this weekend.

Another general observation, I'm wondering if threads in`BasicPager` are the way to go. I wonder if this could be achieved another way, e.i. non-blocking io. To me threads are last resort in implementing things and if I can help I try to stay away.

**Comment by AndrewRadev** [comment]
Here's an example of how it might work using one of the files in the "examples" directory:

``` ruby
$: << File.expand_path('./lib')

require 'tty-pager'

pager = TTY::Pager::SystemPager.new
# pager = TTY::Pager::BasicPager.new

File.readlines('examples/temp.txt').each do |line|
  if !pager.write(line)
    STDERR.puts "Pager was closed"
    break
  end
end

pager.wait
```

I'm not sure if it works quite right, because, at least for me, the pager messes up with the first line. I wonder if the height or width of my terminal is not calculated correctly and if it has to do with my weird linux setup in particular. Will keep experimenting, I guess.

> In other words, it feels to me that there is a common abstraction that links the BasicPager and SystemPager together. Maybe, an enumerable like object that works on newline basis. If you give it a file descriptor it will 'know' how to feed lines to both pagers. Similarly, if you write a line, it will consume it and yield back to the pager. I'm not sure if I'm making much sense.

I considered working with lines as well, but it doesn't feel like it's necessary. If you `write` a chunk of text, that chunk of text can just go to the pager, that's what it boils down to. Really, what I've implemented is a very minimal change to what you already had, in both places -- it was just a matter of separating the `page` methods into three separate parts -- "start" (spawn a process or thread), "write" (which can now be repeated multiple times, which is the big change), and "wait" (for the process or thread to be done). 

The other changes are due to HoundCI, which, to be honest, I'd disable. Some of the suggestions are kind of weird, and even the ones that make sense are to a style that the current codebase doesn't meet. This is going to add style changes to the diff that would probably make sense in a separate PR (though I'd rather you did it, since it's your call which lints you want to disable. My choice would be "most of them" :)).

> Another general observation, I'm wondering if threads inBasicPager are the way to go. I wonder if this could be achieved another way, e.i. non-blocking io. To me threads are last resort in implementing things and if I can help I try to stay away.

Well, the way that I use threads here is pretty basic and uses message-passing with a threadsafe queue rather than shared memory, so I think it's quite sensible in terms of semantics. Performance-wise, I was thinking ruby's threads are green ones rather than os-level, but I realize I'm not actually sure, so that might be unnecessary. I'll look into fibers.

**Comment by AndrewRadev** [comment]
> I'm not sure if it works quite right, because, at least for me, the pager messes up with the first line. I wonder if the height or width of my terminal is not calculated correctly and if it has to do with my weird linux setup in particular. Will keep experimenting, I guess.

In fact, even on master with the `page` method, it breaks for me in the same way, so I definitely think it's something to do with my setup. We'll see.

File: lib/tty/pager/basic.rb [hunk 1]
```diff
@@ -43,50 +45,101 @@ def default_prompt
...
           end
-          output.print(chunk.join)
+          @state.queue << [:print, chunk.join]
 
-          if lines_left == 0
-            break unless continue_paging?(page_num)
-            lines_left = @height
-            if leftover.size > 0
-              lines_left -= leftover.size
+          if @state.lines_left.zero?
```
[the remark this answers is not in the document]
Comment (AndrewRadev) [reply lib/tty/pager/basic.rb:99]: I don't understand what that means and where I should use `next` to skip iteration. Is it the next line that uses `return`? Because I definitely do want to use `return` here, not `next`.

File: lib/tty/pager/basic.rb [hunk 2]
```diff
@@ -43,50 +45,101 @@ def default_prompt
...
+            return false unless continue_paging?
+            @state.lines_left = @height
+            unless @state.leftover.empty?
+              @state.lines_left -= @state.leftover.size
             end
-            page_num += 1
-            return !callback.call(page_num) unless callback.nil?
+            @state.page_num += 1
+            return !callback.call(@state.page_num) unless callback.nil?
```
[the remark this answers is not in the document]
Comment (AndrewRadev) [reply lib/tty/pager/basic.rb:106]: I don't think the extremely marginal gain in performance justifies the weirdness of a line that reads `return !yield`. Is it just me that this looks weird to?

**Comment by AndrewRadev** [comment]
I've now replaced the thread with a fiber, and everything seems to be working fine. It also allowed me to pull the logic into its own `Printer` class, which is nice. 

The more I think about it, though, the more it makes sense to get rid of the `start` method for the basic pager, since the fiber can be initialized in the constructor. It might make sense to do the same for the system pager and encapsulate the "starting" logic in the `PagerIO` class. I'll do some more thinking, but feel free to suggest a direction for this. Having a public `start` method would allow someone to spawn the process before the first print, but that doesn't seem like it'd ever be necessary, and if it is, maybe a `write(nil)` could be made to do just that.

**Comment by AndrewRadev** [comment]
Actually, there's no need to use fibers at all! The way that it works, the fiber simply takes control and blocks when a prompt is needed, but you could achieve the same thing by just printing and calling the `prompt` lambda :D. Really should have seen this earlier, but I guess I was focused on translating the thread setup to a fiber.

What a thread would provide in this case is that the processing logic could keep going and push messages to the queue, and the printer thread would print and block with computation running in the background. But I think that's not really super necessary for the basic printer, so the current setup is probably fine. If you disagree, I could revert the last few commits to get the thread setup back, or do something else if you have another idea.

File: lib/tty/pager/system.rb [hunk 3]
```diff
@@ -123,22 +124,37 @@ def initialize(**options)
...
 
         command = pager_command
         out = self.class.run_command(command)
         # Issue running command, e.g. unsupported flag, fallback to just command
-        if !out.empty?
+        unless out.empty?
           command = pager_command.split.first
         end
 
-        pager_io = open("|#{command}", 'w')
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:135]: The intention behind `tty` gems is to work, if possible, on all systems including Windows. The reason why the `|` pipe is used in this instance, is to make the process piping work on Windows. Have you checked your change against Windows?

File: lib/tty/pager/basic.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (piotrmurach) [reply lib/tty/pager/basic.rb:70]: Don't worry about Hound, probably need to configure hound/rubocop for this project.

File: lib/tty/pager.rb [hunk 4]
```diff
@@ -63,6 +65,15 @@ def enabled?
       !!@enabled
     end
 
+    # Check if pager is currently running
+    #
+    # @return [Boolean]
+    #
+    # @api public
+    def running?
```
Comment (piotrmurach) [inline lib/tty/pager.rb:73]: In what scenario would this be useful?

File: lib/tty/pager/basic.rb [hunk 5]
```diff
@@ -43,50 +45,71 @@ def default_prompt
       #
       # @api public
       def page(text, &callback)
-        page_num = 1
-        leftover = []
-        lines_left = @height
+        write(text, &callback)
```
Comment (piotrmurach) [inline lib/tty/pager/basic.rb:48]: I like the idea of 'preserving' current API. My preference for any updates is to keep the current API working as much as possible.

File: lib/tty/pager/basic.rb [hunk 6]
```diff
@@ -43,50 +45,71 @@ def default_prompt
...
+        reset
+      end
 
+      def write(text, &callback)
         text.lines.each do |line|
           chunk = []
-          if !leftover.empty?
-            chunk = leftover
-            leftover = []
+          unless @state.leftover.empty?
```
Comment (piotrmurach) [inline lib/tty/pager/basic.rb:55]: I'm not sure about putting instance vars together in a state variable. It kind of makes parsing a bit 'too busy'.

File: lib/tty/pager/system.rb [hunk 7]
```diff
@@ -158,6 +174,23 @@ def pager_command(*commands)
                            self.class.find_executable(*commands)
                          end
       end
+
+      class PagerIO
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:178]: I like this, looks like a useful abstraction.

File: lib/tty/pager/system.rb (hunk 3 above)
Comment (AndrewRadev) [reply lib/tty/pager/system.rb:135]: Hm, I admit, I didn't test it on Windows. I didn't know this was the reason. Switching to `IO.popen` had the nice qualities of silencing the Hound warning and making it much more reliable to stub (though I could probably make stubbing just as easy by using `Kernel.open` directly rather than through the class). I'll see what I can do to test it out.

File: lib/tty/pager.rb (hunk 4 above)
Comment (AndrewRadev) [reply lib/tty/pager.rb:73]: I created it to have a single way to check if the pager process was started, so that both `write` and `wait` could respond to the state of the pager. Right now, it's not actually used, because I check for the presence of the `PagerIO`, so I guess I'll remove it.

File: lib/tty/pager/basic.rb (hunk 6 above)
Comment (AndrewRadev) [reply lib/tty/pager/basic.rb:55]: Yeah, I feel like one way to make this nicer would be to do a "law of demeter" pass and create sensible methods to query the state. For instance, `if @state.has_leftover?` or something. Would make this particular clause easier to read, since `unless empty` reads like a double-negative to me.

It was also a way to only check for the presence of a single variable to see if the pager has been started or not. But with the state of things right now, I don't think that would even be necessary. I'll try to turn `State` into a more useful abstraction, but if I don't manage it, I'm happy to roll the pieces back into their separate parts.

File: lib/tty/pager/system.rb (hunk 3 above)
Comment (piotrmurach) [reply lib/tty/pager/system.rb:135]: I'm all for switching over to `popen` if it makes things nicer but would like to keep the windows support as well. Maybe we can have both with this. I just don't know without testing. I picked up this particular 'trick' from Avdi Grimm and it reduced my 'hacky' code to this wonderful one character.

File: lib/tty/pager/basic.rb (hunk 6 above)
Comment (piotrmurach) [reply lib/tty/pager/basic.rb:55]: I like the idea of `reset` and I'd probably keep all the mutating vars in there for now. In my experience, this simplifies the logic. I'd be all for extraction of the `State` if it attracted some common behaviour etc... In this instance, it feels unnecessary.

File: lib/tty/pager/basic.rb [hunk 8]
```diff
@@ -43,50 +45,71 @@ def default_prompt
...
+            chunk = @state.leftover
+            @state.leftover = []
           end
           wrapped_line = Strings.wrap(line, @width)
           wrapped_line.lines.each do |line_part|
-            if lines_left > 0
+            if @state.lines_left.positive?
```
Comment (piotrmurach) [inline lib/tty/pager/basic.rb:61]: I know this is pain, but I'd keep the `> 0` comparison. Why? At the moment this gem is meant to work with Ruby `>= 2.0`. Sorry!

File: lib/tty/pager/basic.rb (hunk 6 above)
Comment (piotrmurach) [reply lib/tty/pager/basic.rb:55]: Sorry, it may be a bit 'cheeky' of me to say but the fact that this is named `State` indicates that it doesn't have a clear role/purpose. For example, if it was named `PageView` or something similar, it would give a bit more context to its existence.

Comment (AndrewRadev) [reply lib/tty/pager/basic.rb]: It's a valid point -- the name "State" doesn't really encapsulate any domain logic, it's essentially a value struct that hold properties in one place.

I've pulled it out into a `Pagination` class that I think does a better job of encapsulating the domain logic of chunking the given lines into pages. It could even be initialized with the `@width` as well to handle the wrapping part, but I think it's fine like this. It's not reusable, but I think the separation makes the iteration and pagination a bit easier to read. It could also be tested separately if it were pulled out into another file, although I'm not sure if I'd go in that direction.

If you still think the separation adds a level of indirection that could be confusing or add overhead when reading, let me know and I'll revert to using the individual variables.

File: lib/tty/pager/basic.rb (hunk 8 above)
Comment (AndrewRadev) [reply lib/tty/pager/basic.rb]: Yeah, I'm happy to use it as well. I think I'll just go ahead and ignore future Hound complaints :).

File: lib/tty/pager/system.rb (hunk 3 above)
Comment (AndrewRadev) [reply lib/tty/pager/system.rb:135]: Okay, so I tested it on Windows 10, and it seems to work just fine:
tty-prompt

I'm not 100% sure if this will work on all setups, since ruby and git install a bunch of development stuff, but I also feel like `Kernel.open` probably just delegates to this anyway? The problem with it is it's also used to open files and URIs, so `popen` is clearer in terms of intent, if nothing else. I'd keep it for now.

Comment (piotrmurach) [reply lib/tty/pager/system.rb:135]: That looks promising. Thanks for checking Windows. I normally test on XP VM - a bit old school I know. I don't know the internals of `popen` but semantically it seems to be doing what the pipe character used to do - open a child process and pipe to it. Let's go with `popen` for now.

File: lib/tty/pager/basic.rb (hunk 6 above)
Comment (piotrmurach) [reply lib/tty/pager/basic.rb]: I like what you've done here - the `Pagination` seems like more useful abstraction. Having said that, my thinking about this spike is to get the manual page chunks writing done first for all the pagers. From the point of what we're aiming to do this is a 'design improvement' diversion. I'd be inclined to skip it for now and first to do the `write` call as we want/need it. I think this spike has great stuff in it and I'd like to get to the point with it where I'm good to merge some implementation, whether you rebase later or submit another PR. Everything else can be done in separate PRs later. To me the smaller and focused PR the better. Plus I think that the basic pager line processing is a bit broken abstraction anyhow. It worked for me at the time. But now, it will fall down at the first bit when someone wants to page a continuous blob of text. So in other words, we may need to consider a different way of text processing for the basic pager anyhow. What do you think?

File: spec/unit/null/page_spec.rb [hunk 9]
```diff
@@ -20,4 +20,14 @@
...
+
+  it "doesn't write a newline if one wasn't given" do
+    allow(output).to receive(:tty?).and_return(true)
+    pager = described_class.new(output: output)
+    text = "I try all things, I achieve what I can."
+
+    pager.page(text)
+
+    expect(output.string).to eq(text)
+  end
```
Comment (AndrewRadev) [inline spec/unit/null/page_spec.rb:32]: The reason for this change is because the `page` method doesn't use `puts` now, it uses `write`. Why is this different? Because (and this was kind of a surprise to me), if you `puts "something\n"` and `puts "something"`, those two produce the same results. So, `puts` doesn't *add* a newline, it ensures there is one at the end. I'm not sure if this was intentional, but there wasn't a test for this edge case, so here it is. It can be changed to "it writes a newline even if one wasn't given" and the code for `page` could use `puts` instead.

**Comment by AndrewRadev** [comment]
I've added documentation as best as I could, and an explanation in the README. I've also filled in some missing tests. Coverage is still down, but the methods that are being invoked are extremely simple delegations... Like, I don't think there's much to test for the method `wait` that returns `true` for the `NullPager`. Obviously, I could still add a test to fill in coverage. I'll follow your judgement on this.

The same thing for HoundCI -- let me know which style fixes you'd like me to apply.

The TravisCI build fails on jruby, but I'm not sure why. The backtrace doesn't seem like anything that comes from the test suite.

> From the point of what we're aiming to do this is a 'design improvement' diversion. I'd be inclined to skip it for now and first to do the write call as we want/need it. I think this spike has great stuff in it and I'd like to get to the point with it where I'm good to merge some implementation, whether you rebase later or submit another PR. Everything else can be done in separate PRs later. To me the smaller and focused PR the better.

Okay, I've put back the original logic, except with instance vars instead of local vars. It works.

> It worked for me at the time. But now, it will fall down at the first bit when someone wants to page a continuous blob of text. So in other words, we may need to consider a different way of text processing for the basic pager anyhow.

Hm, well, I don't know about that. It seems like it works fine for me, what scenario do you see that it would break in? Or do you just mean its interface is too basic?

In any case, let me know if you'd like me to merge this, squash and rebase, change documentation, naming etc. I could make a new PR as well to have a clean slate (would be particularly useful if you disable HoundCI, so we can see reduce the noise in the new PR :D). Let me know how you'd like to proceed.

File: README.md [hunk 10]
```diff
@@ -56,6 +56,28 @@ Then to perform actual content pagination invoke `page` method with the content
...
+
+  if success
+    # keep going to the next line
+  else
+    # pager was closed, stop writing
+    break
+  end
+end
+
+# wait for the interface to be closed
```
Comment (piotrmurach) [inline README.md:77]: I'm not sure about the 'interface' wording if it is clear enough. What do you think about - wait for the paging utility to be closed?

File: lib/tty/pager/system.rb [hunk 11]
```diff
@@ -158,6 +194,26 @@ def pager_command(*commands)
...
+        def initialize(command)
+          @io  = IO.popen(command, 'w')
+          @pid = @io.pid
+        end
+
+        def write(*args)
+          @io.write(*args)
+        end
+
+        def close
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:211]: It seems like this method could benefit from splitting into `close` and `wait`.

File: lib/tty/pager/system.rb [hunk 12]
```diff
@@ -123,22 +124,57 @@ def initialize(**options)
...
+      # Stop the pager, wait for the process to finish. If no pager has been
+      # started, returns true.
+      #
+      # @return [Boolean]
+      #   the exit status of the child process
+      #
+      # @api public
+      def wait
+        return true unless @pager_io
+        status = @pager_io.close
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:176]: I'd be inclined to split `@pager_io.close` into `@pager_io.close` and `@pager_io.wait` and move the `Errno::ECHILD` to the `PagerIO` to hide this detail.

File: lib/tty/pager/system.rb [hunk 13]
```diff
@@ -123,22 +124,57 @@ def initialize(**options)
...
+        wait
+      end
+
+      # Spawn the pager process
+      #
+      # @return [PagerIO]
+      #   A wrapper for the external pager
+      #
+      # @api private
+      def start
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:137]: I like your comment and since this is a private API we could provide more intention revealing name like `spawn_pager`?

File: lib/tty/pager/system.rb [hunk 14]
```diff
@@ -123,22 +124,57 @@ def initialize(**options)
...
+      #
+      # @return [Boolean]
+      #   the success status of writing to the pager process
+      #
+      # @api public
+      def write(text)
+        @pager_io ||= start
+        @pager_io.write(text)
+        true
+      rescue Errno::EPIPE
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:162]: I'd hide the error inside the `@pager_io`, what do you think?

File: lib/tty/pager/system.rb [hunk 15]
```diff
@@ -123,22 +124,57 @@ def initialize(**options)
...
-        pager_io.write(text)
-        pager_io.close
+      # Send text to the pager process. Starts a new process if it hasn't been
+      # started yet. Returns false if the pager was closed.
+      #
+      # @return [Boolean]
+      #   the success status of writing to the pager process
+      #
+      # @api public
+      def write(text)
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:158]: Couple things to make the API more flexible. First, `write` could accept any number of strings? Second, we could add `#<<` alias to mimic IO writing?

File: spec/unit/null/page_spec.rb (hunk 9 above)
Comment (piotrmurach) [reply spec/unit/null/page_spec.rb:32]: I agree with your observation. Equally, why don't we add `puts` as a close sibling to `write` to the pagers?

File: spec/unit/system/wait_spec.rb [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  it "succeeds if the pager exits successfully" do
+    pager = described_class.new
+    pager_io = double("PagerIO", write: nil, close: double(success?: true))
+    expect(pager).to receive(:start).once.and_return(pager_io)
+
+    pager.write("test")
+
+    expect(pager.wait).to be_truthy
```
Comment (piotrmurach) [inline spec/unit/system/wait_spec.rb:18]: I tend to prefer a direct value comparison `eq(true)` as I was bitten by the truthy matcher a few times in the past. There are too many things that can be truthy in Ruby.

File: README.md (hunk 10 above)
Comment (AndrewRadev) [reply README.md]: I was going for "the user interface" I guess, but I can see how it's confusing. I changed it to "the paging tool", how does that sound to you? I was thinking of "the pager process", but that's not quite right for the basic pager.

File: lib/tty/pager/system.rb (hunk 12 above)
Comment (AndrewRadev) [reply lib/tty/pager/system.rb]: Yes, encapsulating IO details makes sense :+1:

File: lib/tty/pager/system.rb (hunk 13 above)
Comment (AndrewRadev) [reply lib/tty/pager/system.rb]: Agreed, that's a much better name.

File: lib/tty/pager/system.rb (hunk 15 above)
Comment (AndrewRadev) [reply lib/tty/pager/system.rb:158]: Hm, I didn't know the built-in `write` accepts multiple strings. It's a bit weird to me since it seems like you could always concatenate them, but I suppose it would be less efficient. I'll probably do that tomorrow.

I implemented `<<` as an alias, but I'm not sure it's a great idea. The problem is that `pager << "foo" << "bar"` could only work if that method returns `self`, but that method currently returns true or false to indicate if the pager was closed. Instead of a boolean result, it could raise an error, but it would make handling that case a bit more annoying.

File: lib/tty/pager/system.rb [hunk 17]
```diff
@@ -123,26 +124,69 @@ def initialize(**options)
...
+      end
+
+      # Stop the pager, wait for the process to finish. If no pager has been
+      # started, returns true.
+      #
+      # @return [Boolean]
+      #   the exit status of the child process
+      #
+      # @api public
+      def wait
```
Comment (piotrmurach) [inline lib/tty/pager/system.rb:184]: I will come clean here. I'm not totally sold on `wait` method. Why? This name feels like an implementation detail. Does a user need to know that they are waiting for a process?  The name breaks down a little for me in the basic pager implementation. Because of this, I'm leaning towards the `close` method. Fundamentally, this is the intention, open a pager, keep on writing and when you done close it. This also happens to mimick the `File`.

```ruby
pager = TTY::Pager.new
pager << "I know not all that may be coming"
pager << "but be it what it will, I'll go to it laughing."
pager.close
```
If user doesn't `close` the pager the expectation is that it will remain waiting. Similarly, we could allow for class-level method like `page` or `open` with a block that would close things automatically for you:

```ruby
TTY::Pager.page(command: "less -R") do |pager|
  pager << "I know not all that may be coming"
  pager << "but be it what it will, I'll go to it laughing."
end
```

Then we could allow for paging an entire file with one line:

```ruby
TTY::Pager.page(File.open("large-file.txt"), command: "less -R")
```

I think there is a significant benefit to making things similar/familiar to what Rubyists may already experience. Any thoughts?

Comment (AndrewRadev) [reply lib/tty/pager/system.rb]: Yes, this makes perfect sense. Considering the need to start and wait, a block form makes a lot more sense as the "canonical" way to interact with the pager. And `close` is definitely a better name.

With that, and the write and puts aliases, I'm thinking more and more that instead of a boolean result, an exception would be much better to indicate that the pager has been closed. Particularly since it means the block form could just handle it automatically in a way that a `return false` couldn't.

So, I've implemented that, adding a `try_write` method for finer-grained control. Although I don't think it's that necessary -- I'd be happy to remove it. I could also try using `throw` instead of `raise` since it's debatable whether the pager process being closed is an "error" (the process does go away while the program is working, after all), or "control flow" (it's an expected thing that could happen because of user interaction). For all the dislike of using exceptions as control flow, they're slightly more familiar to users, so ¯\_(ツ)_/¯.

Let me know what you think. 

I've yet to write some tests for the block form. I'll get around to it. One thing I want to mention as an aside is that I feel that separating unit tests by methods per file feels a bit limiting. Multiple methods could easily be tested in one example, like `write` and `puts`, which now duplicate quite a lot of tests. But this is my own opinion, and I'm happy to leave things as they are. As you say, probably not the best idea to make too many changes in one PR.

Comment (AndrewRadev) [reply lib/tty/pager/system.rb]: The current "lorem" example:

``` ruby
lorem = <<~EOF
  ... lots of text ...
EOF

TTY::Pager.page do |pager|
  10.times do
    pager.write(lorem)
    sleep 1
    pager.write "---------------------\n"
  end
end
```

File: lib/tty/pager/basic.rb [hunk 18]
```diff
@@ -43,48 +45,112 @@ def default_prompt
...
+          if @lines_left == 0
+            raise PagerClosed unless continue_paging?(@page_num)
+            @lines_left = @height
+            if @leftover.size > 0
+              @lines_left -= @leftover.size
             end
-            page_num += 1
-            return !callback.call(page_num) unless callback.nil?
+            @page_num += 1
+            callback.call(@page_num) unless callback.nil?
```
Comment (AndrewRadev) [inline lib/tty/pager/basic.rb:126]: I'm not sure what the idea behind this callback is. I guess there should be a `return` here to exit the function, but I'm not sure about the conditions. What's the case for invoking `write` with a callback?

File: lib/tty/pager/system.rb [hunk 19]
```diff
@@ -158,6 +227,41 @@ def pager_command(*commands)
...
+        def write(*args)
+          io_call(:write, *args)
+        end
+
+        def puts(*args)
+          io_call(:puts, *args)
+        end
+
+        def close
+          @io.close
```
[the remark this answers is not in the document]
Comment (AndrewRadev) [reply lib/tty/pager/system.rb:249]: Testing it out, calling `close` twice doesn't seem to error out, but I'm not sure if this is cross-platform behaviour, so I'll add a check.

It would be nice to write a test for it, but there wouldn't be a point if it uses stubs. I feel like this project could benefit a lot from integration tests with real process-spawning, but that's definitely not in the scope of this PR.

[the remark this answers is not in the document]
Comment (piotrmurach) [reply lib/tty/pager/system.rb:251]: 💯 agree on integration tests and possibly performance tests as well with [rspec-benchmark](https://github.com/piotrmurach/rspec-benchmark) . I wrote things quite a while back and definitely the test suite needs more love.

File: lib/tty/pager/basic.rb [hunk 20]
```diff
@@ -43,48 +45,112 @@ def default_prompt
...
-          output.print(chunk.join)
+          output.public_send(write_method, chunk.join)
 
-          if lines_left == 0
-            break unless continue_paging?(page_num)
-            lines_left = @height
-            if leftover.size > 0
-              lines_left -= leftover.size
+          if @lines_left == 0
+            raise PagerClosed unless continue_paging?(@page_num)
```
[the remark this answers is not in the document]
Comment (AndrewRadev) [reply lib/tty/pager/basic.rb]: I've added a more readable error message, and I've made it specific to the pager, with more information for debugging purposes. Let me know what you think.

Regarding the "closed" variable, I'm not sure how it would work to avoid the `try_write`. I'm imagining something like

``` ruby
if !pager.closed?
  pager.write(...)
end
```

But we can't have an instance variable for that -- the act of writing is the thing that discovers that the stream is closed, and that happens after the check. We could check the `closed?` state of the pipe, but then there could be a timing issue with the check succeeding, but then in the middle of the `write` call the user closes the pager. It's probably unlikely, it just feels a bit flaky to me. Let me know if I'm misunderstanding the scenario, though.

File: lib/tty/pager/system.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (AndrewRadev) [reply lib/tty/pager/system.rb:251]: I've done integration tests for process-juggling before, and I was reasonably happy with how it turned out: [waiting_on_rails](https://github.com/AndrewRadev/waiting-on-rails/blob/master/spec/waiting_on_rails/rails_spec.rb). Essentially, the main idea was stubbing the `PATH` to ensure a fake `rails` command and a fake `mplayer` command get called, and those would be driven by the test suite via sockets. This is probably quite non-portable, though. Still, there are options.

File: lib/tty/pager/basic.rb (hunk 20 above)
[the remark this answers is not in the document]
Comment (piotrmurach) [reply lib/tty/pager/basic.rb]: I agree that's probably not a substitute for the `try_wite` in which case it makes sense to keep it. The `closed?` relates more to the process/pipe dying. When the paging tool died then 100% you cannot make any writes. And this probably works better with `<<` to provide an extra layer of robustness. However, this won't mitigate the scenario when the failure happens during the write - then all bets are off. In which case raising an error is a reasonable approach or calling `try_write`.

File: lib/tty/pager.rb [hunk 21]
```diff
@@ -73,11 +86,51 @@ def enabled?
...
 
+    # Close the pager tool, wait for it to finish
+    #
+    # @return [Boolean]
+    #   whether the pager exited successfully
+    #
+    # @api public
+    def close
+      pager.close
+    end
```
Comment (piotrmurach) [inline lib/tty/pager.rb:132]: How about we use `forwardable` module to delegate these?

File: lib/tty/pager/basic.rb [hunk 22]
```diff
@@ -42,49 +44,115 @@ def default_prompt
...
+      # @return [TTY::Pager::BasicPager]
+      #
+      # @api public
+      def write(*args)
+        args.each do |text|
+          send_text(:write, text)
+        end
+        self
+      end
+      alias_method :<<, :write
```
Comment (piotrmurach) [inline lib/tty/pager/basic.rb:69]: In my experience `alias` is usually more desirable.

File: lib/tty/pager.rb (hunk 21 above)
Comment (AndrewRadev) [reply lib/tty/pager.rb:132]: This is doable for `close`, although I do get this warning in tests: 

``` ruby
warning: TTY::Pager#close at /home/andrew/projects/tty-pager/lib/tty/pager.rb:132 forwarding to private method RSpec::Mocks::Double#close
```

The reason this happens is because the `pager` is a protected attribute, I think. Though replacing the `spy(:system_pager)` with a `spy(:system_pager, close: true)` no longer shows the error, so I'm not sure what the deal is.

For the other methods like `puts` and `write`, it's a bit trickier, because their interface is supposed to return `self`, but if we delegate, the `self` is going to be the inner pager, not the container. Which seems subtly counter-intuitive. It shouldn't be a practical problem, but it just means that `pager << "foo" << "bar"` changes its type halfway through, which is kind of weird.

To be honest, the relationship between `TTY::Pager` and the other ones is a bit weird. It's both their parent class, and also a container for pager instances. I'd consider separating their responsibilities -- have the `NullPager` be a parent, or some form of an abstract pager that has all the methods implemented but raises errors to ensure all the children implement it. Have the `TTY::Pager` class have `.page` with a block and `.get_pager(**options)` or something, and nothing else. But that's something for later.

Comment (piotrmurach) [reply lib/tty/pager.rb]: Ahh I see - never mind. Sorry for the trouble. Let's forget about my suggestion and revert back to what it was. Introducing a module for only one method doesn't seem to be worth it. 

Agreed, an interface/abstract class for all the three pager types would be good. This is what happens when you bolt-on new pagers over time and don't stop to refactor. It all started with a system pager only and then grew branches. Definitely would be interested in making this cleaner and potentially more extensible for when people want to implement their own paging logic or extend existing. However, this kind of change would need to wait for the next release. I'd rather have many releases than one big bang release. This will allow for feedback from potential users.

**Comment by AndrewRadev** [comment]
> Apart from that, I'm happy for you to submit a new PR and let's finally merge this! 

I've submitted https://github.com/piotrmurach/tty-pager/pull/15 with squashed commits. Let me know if this is what you had in mind, or would prefer something different.
